### ISSUE TITLE
【RAG】RAG 入门

### ISSUE THREAD

**Issue by QingyaFan** [QingyaFan/blog#92]
# 什么是 RAG 技术

RAG 是 Retrieval-Argumented Generation，检索增强生成，增强的是LLM的生成能力，是对LLM知识固化过时的增强，而增强的措施就是检索。那么RAG 技术中就存在三个要素：知识、检索、生成。
- 知识，可能是文档，如沉淀的用户文档，SOP等；或者一些实时的数据，如系统状态，可以是Metric、Log、Trace等实时产生的数据。当然，它们得处理成大模型能懂的表示：embeddings（向量）
- 检索，检索需要支持向量的距离算法，来近似语义的相似度
- 生成，检索的结果交给 LLM 来总结生成。

## RAG 与 微调（Fine Tune）

RAG 经常与微调放在一起讨论，它们可以达成类似的效果。但它们是不同的技术，有不同的特点，需要根据实际情况灵活选用：

|特性 | RAG | 微调|
|-- | -- | --|
|适用场景 | 知识频繁更新、多领域知识集成 | 明确任务、特定领域高精度需求|
|知识更新 | 知识库可随时更新，实时反映 | 需重新微调模型，而微调需要高质量的标注数据|
|计算成本 | 低（无需重新训练模型） | 高（需大量计算资源进行微调）|
|灵活性 | 高（适应多领域、动态任务） | 低（特定领域优化，通用性下降）|
|性能 | 依赖检索和知识库质量 | 在特定任务上性能更优|

## Agentic RAG

Agentic RAG 到底是 RAG 还是 Agent，业界没有统一的看法。例如LlamaIndex认为是RAG [^13]，LangChain认为是 Agent。[^12]

# 实现一个简易RAG应用

组成部分：

典型的 RAG 应用包括 `离线数据处理` 和 `在线检索生成`两部分：
1. 离线数据处理（indexing）：从数据源获取数据，建立索引（index），一般发生在离线阶段。
2. 在线检索生成（Retrieval and Generation）：RAG 的主体部分，与用户交互获取`user query`，从索引检索相关数据，交给大模型总结。

### 数据处理

数据质量很关键：
- 格式：markdown 格式属于半结构化数据，包含标题、表格、图片等丰富信息，便于进行语义切割；且大模型的预训练中有丰富的markdown格式数据，支持较好。
- 清洗，todo
- chunking：对内容进行合理分割
- embedding：

#### 为什么对数据进行切分？

todo

### 选定的 Demo 技术栈

一些流行的库如下：
- langchain
- llamaindex
- haystack
- ragflow

技术栈：我们可以使用 open-webui、pgvector、LangChain/LlamaIndex 组合的技术栈：
- pgvector存储文档转换成的向量
- LangChain/LlamaIndex负责
    1. 调用embedding大模型，将文档转换成向量格式，存入pgvector
    2. 从pgvector检索结果，将结果交由generation大模型总结
- open-webui负责与用户交互

## 使用LangChain

文档或知识分为 unstructured 和 structured，一个产品的文档是 unstructured，SQL 知识是structured。

### 离线部分

在LangChain中，将文档转换存储到矢量存储的过程分为三个步骤：load、split、store：

#### load

从磁盘中读取文档内容，并为其建立metadata；

#### split

为什么要split？一个原因是大模型有 max tokens 限制，如果一个文档较长，不split很难放到现有模型里。被split的每个部分的大小就要因文档而异了，且为了文档之间的上下文之间能建立联系，可以选择重叠长度。一般来说 1000的chunk长度，200的重叠长度比较中庸。[^10] 如下：

```
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=1000, chunk_overlap=200
)
```

#### store

【todo】

### 在线检索

在 LangChain中，Retriever 对象负责接收query返回相关documents。

todo

#### 记住会话历史

LangGraph 实现了一个内置的 [persistence layer](https://langchain-ai.github.io/langgraph/concepts/persistence/)，适合用于有多轮对话的应用。

doing

### 设置 pgvector

```
CREATE EXTENSION vector;

CREATE TABLE documents (id bigserial PRIMARY KEY, text text, embeddings vector(1536))
```

#### 将文档导入 pgvector

首先要将文档转换成embeddings，使用LlamaIndex来加载并存储文档到pgvector。

```
pip install llama-index openai psycopg2
```

解析并存储

```
from PyPDF2 import PdfReader
from llama_index.embeddings.openai import OpenAIEmbedding
import psycopg2


def extract_text_from_pdf(pdf_path):
    reader = PdfReader(pdf_path)
    text = ""
    for page in reader.pages:
        text += page.extract_text()
    return text

def generate_embeddings(text):
    embedding_model = OpenAIEmbedding()
    vector = embedding_model.get_text_embedding(text)
    return vector

def store_embedding_in_pgvector(text, embedding, connection_string):
    conn = psycopg2.connect(connection_string)
    cur = conn.cursor()

    cur.execute(
        "INSERT INTO documents (text, embedding) VALUES (%s, %s)",
        (text, embedding)
    )

    conn.commit()
    cur.close()
    conn.close()

pdf_path = "example.pdf"
connection_string = "dbname=mydb user=me password=ps host=h"

text = extract_text_from_pdf(pdf_path)
embedding = generate_embeddings(text)
store_embedding_in_pgvector(text, embedding, connection_string)
```

#### 使用pgvector中的文档

使用LlamaIndex来使用pgvector中的数据。[^7]

```
# pip install llama-index-vector-stores-postgres
# pip install llama-index
# pip install SQLAlchemy

from sqlalchemy import make_url
from llama_index.vector_stores.postgres import PGVectorStore
from llama_index.core import StorageContext, VectorStoreIndex, SimpleDirectoryReader

documents = SimpleDirectoryReader("./data").load_data()

url = make_url("dsn")
vector_store = PGVectorStore.from_params(
    database="db_name",
    host=url.host,
    password=url.password,
    port=url.port,
    user=url.username,
    table_name="paul_graham_essay",
    embed_dim=1536, # change by the embedding model
    hnsw_kwargs={
        "hnsw_m": 16,
        "hnsw_ef_construction": 64,
        "hnsw_ef_search": 40,
        "hnsw_dist_method": "vector_cosine_ops",
    }
)
storage_context = StorageContext.from_defaults(vector_store=vector_store)
index = VectorStoreIndex.from_documents(
    documents, storage_context=storage_context, show_progress=True
)
query_engine = index.as_query_engine()
```

### 使用 open-webui

启动一个open-webui实例。

```
docker run -d \
    -p 3000:8080 \
    -v open-webui:/app/backend/data \
    -e WEBUI_AUTH=False \
    --name open-webui \
    ghcr.io/open-webui/open-webui:main
```

todo: 为什么需要挂volume？oi不是单纯的web ui吗？

### 配置使用OpenAI兼容的API

这里比较自由，我们可以使用ollama或vLLM启动的LLM服务。

使用 ollama 服务：

```
from llama_index.embeddings.ollama import OllamaEmbedding

ollama_embedding = OllamaEmbedding(
    model_name="xxxx",
    base_url="xxxxxxx",
)
```

使用 vLLM 服务：[^9]

```
from llama_index.llms.vllm import VllmServer
from llama_index.core.llms import ChatMessage

llm = VllmServer(
    api_url="xxxx",
    temperature=0,
    max_new_tokens=100,
)

llm.complete("xxx")
```

temperature 参数用于控制LLM 生成文本时的随机性和多样性。具体来说，temperature 决定了模型在生成下一个词时的概率分布的平滑程度，从而影响生成文本的“创造性”或“确定性”。其原理如下：
- 在生成文本时，模型会为每个可能的下一个词分配一个概率。这些概率通过对模型的输出 logits 应用 softmax 函数得到。temperature 参数用于调整这些 logits，公式：`softmax(logits / temperature)`。当 temperature 较低（接近于0）时，模型会更加确定性地选择概率最高的词，生成的文本更为一致和可预测；当 temperature 较高时，概率分布会变得更加平坦，模型在生成文本时会有更多的随机性，可能生成更多样化和富有创意的内容。所以，**低温度（例如 temperature=0）适用于需要高准确性和一致性的任务，如代码生成、问答系统等；高温度（例如 temperature=0.7 或更高）适用于需要创造性和多样性的任务，如写作、故事生成、头脑风暴等。**
    - softmax 是什么？softmax 函数是一种常用的激活函数，主要用于多分类任务的输出层。它的主要作用是将模型的输出（通常成为“logits”或未归一化的分数）转换为概率分布，使得每个类别的输出值介于 0~1 之间，并且所有类别的概率之和为 1。这种转换有助于模型在多分类问题中进行概率化的预测。

max_new_tokens 参数的作用是限制LLM 在生成回答时最多可以生成的新标记（tokens）的数量。也就是说设置 `max_new_tokens=100` 后，无论 prompt 有多长，响应结果都不会超过 100 tokens。**但要注意：这个限制可能会导致回答不完整，被截断。** 一般用途如下：
- 控制输出长度：适用于需要简洁回答的应用场景，如聊天机器人、摘要生成等；
- 资源管理：限制生成标记数量可以节省计算资源和内存，特别是在处理大型模型或在资源受限的环境中运行时；
- 防止过度生成：避免模型因无限制生成而产生冗长或不相关的内容，从而提高响应的质量和相关性。

[^1]: https://colab.research.google.com/github/run-llama/llama_index/blob/main/docs/docs/examples/vector_stores/postgres.ipynb#scrollTo=8731da62
[^2]: https://docs.llamaindex.ai/en/stable/optimizing/production_rag/
[^7]: https://docs.llamaindex.ai/en/stable/examples/vector_stores/postgres/
[^8]: https://github.com/pgvector/pgvector?tab=readme-ov-file#querying
[^9]: https://docs.llamaindex.ai/en/latest/examples/llm/vllm/
[^10]: https://langchain-ai.github.io/langgraph/tutorials/rag/langgraph_agentic_rag/
[^13]: https://www.llamaindex.ai/blog/agentic-rag-with-llamaindex-2721b8a49ff6
